third_party/xla_client/thread_pool.cc

```diff
@@ -1,18 +1,180 @@
 #include "tensorflow/compiler/xla/xla_client/thread_pool.h"
 
-#include <thread>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <mutex>
+
+#include "tensorflow/compiler/xla/xla_client/metrics.h"
+#include "tensorflow/compiler/xla/xla_client/tf_logging.h"
 
 namespace xla {
 namespace env {
+namespace {
+
+class ThreadPool {
+ public:
+  explicit ThreadPool(size_t num_threads) {
+    threads_.reserve(num_threads);
+    for (size_t i = 0; i < num_threads; ++i) {
+      threads_.emplace_back([this]() { Worker(); });
+    }
+  }
+
+  ~ThreadPool() {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      done_ = true;
+      cv_.notify_all();
+    }
+    for (auto& thread : threads_) {
+      thread.join();
+    }
+  }
+
+  void Schedule(std::function<void()> closure) {
+    // If we have more work scheduled than waiting worker threads, just schedule
+    // it on a separate thread. This prevents tricky thread-pool-size-deadlocks
+    // caused by an undersized thread pool and closures that end up doing sync
+    // waits on the pool threads.
+    bool scheduled = false;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (work_.size() < waiting_) {
+        work_.emplace_back(std::move(closure));
+        scheduled = true;
+      }
+    }
+    if (scheduled) {
+      cv_.notify_one();
+    } else {
+      ScheduleOnThread(std::move(closure));
+    }
+  }
+
+ private:
+  void Worker() {
+    while (true) {
+      std::function<void()> closure = GetWork();
+      if (closure == nullptr) {
+        break;
+      }
+      try {
+        closure();
+      } catch (const std::exception& ex) {
+        XLA_COUNTER("ThreadPoolException", 1);
+        TF_LOG(ERROR) << "Exception from running thread pool closure: "
+                      << ex.what();
+      }
+    }
+  }
+
+  void ScheduleOnThread(std::function<void()> closure) {
+    std::thread thread(std::move(closure));
+    thread.detach();
+  }
+
+  std::function<void()> GetWork() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    ++waiting_;
+    cv_.wait(lock, [this] { return done_ || !work_.empty(); });
+    --waiting_;
+    if (work_.empty()) {
+      return nullptr;
+    }
+    std::function<void()> closure(std::move(work_.front()));
+    work_.pop_front();
+    return closure;
+  }
+
+  std::vector<std::thread> threads_;
+  std::mutex mutex_;
+  std::condition_variable cv_;
+  bool done_ = false;
```

Comment:
I would add comments for these fields, especially for this one.

Reply:
Isn't the naming and the code obvious enough. It's the most boring thread pool ever conceived :)
I really struggle in what I should comment. This is mutex/condition_variable 101.